File: src/searching/cpp/interpolation_search/test/test_interpolation_search.cpp
/**
 * @file test_interpolation_search.cpp
 * @brief Unit tests for the interpolation search algorithm.
 */

#include "interpolation_search.h"
#include "gtest/gtest.h"

/// @brief Test searching an empty array.
TEST(InterpolationSearchTest, EmptyArray) {
  std::vector<int> arr;
  EXPECT_EQ(interpolation_search(arr, 5), std::nullopt);
}

/// @brief Test finding the only element in a single-element array.
TEST(InterpolationSearchTest, SingleElementFound) {
  std::vector<int> arr{5};
  EXPECT_EQ(interpolation_search(arr, 5), 0);
}

/// @brief Test not finding element in a single-element array.
TEST(InterpolationSearchTest, SingleElementNotFound) {
  std::vector<int> arr{5};
  EXPECT_EQ(interpolation_search(arr, 3), std::nullopt);
}

/// @brief Test target at the first position.
TEST(InterpolationSearchTest, TargetAtBeginning) {
  std::vector<int> arr{10, 20, 30, 40, 50};
  EXPECT_EQ(interpolation_search(arr, 10), 0);
}

/// @brief Test target at the last position.
TEST(InterpolationSearchTest, TargetAtEnd) {
  std::vector<int> arr{10, 20, 30, 40, 50};
  EXPECT_EQ(interpolation_search(arr, 50), 4);
}

/// @brief Test target in the middle of the array.
TEST(InterpolationSearchTest, TargetInMiddle) {
  std::vector<int> arr{10, 20, 30, 40, 50};
  EXPECT_EQ(interpolation_search(arr, 30), 2);
}

/// @brief Test target smaller than all elements.
TEST(InterpolationSearchTest, TargetTooSmall) {
  std::vector<int> arr{10, 20, 30, 40, 50};
  EXPECT_EQ(interpolation_search(arr, 5), std::nullopt);
}

/// @brief Test target larger than all elements.
TEST(InterpolationSearchTest, TargetTooLarge) {
  std::vector<int> arr{10, 20, 30, 40, 50};
  EXPECT_EQ(interpolation_search(arr, 60), std::nullopt);
}

/// @brief Test target in gap between elements.
TEST(InterpolationSearchTest, TargetInGap) {
  std::vector<int> arr{10, 20, 30, 40, 50};
  EXPECT_EQ(interpolation_search(arr, 25), std::nullopt);
}

/// @brief Test array with all same elements, target matches.
TEST(InterpolationSearchTest, AllSameElementsFound) {
  std::vector<int> arr{5, 5, 5, 5};
  EXPECT_EQ(interpolation_search(arr, 5), 0);
}

/// @brief Test array with all same elements, target doesn't match.
TEST(InterpolationSearchTest, AllSameElementsNotFound) {
  std::vector<int> arr{5, 5, 5, 5};
  EXPECT_EQ(interpolation_search(arr, 3), std::nullopt);
}

File: src/searching/python/interpolation_search/tests/test_interpolation_search.py
"""Tests for interpolation search algorithm."""

import unittest
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from interpolation_search import interpolation_search


class TestInterpolationSearch(unittest.TestCase):
    """Test cases for the interpolation_search function."""

    def test_empty_array(self):
        """Searching an empty array returns None."""
        self.assertIsNone(interpolation_search([], 5))

    def test_single_element_found(self):
        """Searching for the only element in a single-element array."""
        self.assertEqual(interpolation_search([5], 5), 0)

    def test_single_element_not_found(self):
        """Searching for a missing element in a single-element array."""
        self.assertIsNone(interpolation_search([5], 3))

    def test_target_at_beginning(self):
        """Target is at the first position."""
        self.assertEqual(interpolation_search([10, 20, 30, 40, 50], 10), 0)

    def test_target_at_end(self):
        """Target is at the last position."""
        self.assertEqual(interpolation_search([10, 20, 30, 40, 50], 50), 4)

    def test_target_in_middle(self):
        """Target is in the middle of the array."""
        self.assertEqual(interpolation_search([10, 20, 30, 40, 50], 30), 2)

    def test_target_not_found_too_small(self):
        """Target is smaller than all elements."""
        self.assertIsNone(interpolation_search([10, 20, 30, 40, 50], 5))

    def test_target_not_found_too_large(self):
        """Target is larger than all elements."""
        self.assertIsNone(interpolation_search([10, 20, 30, 40, 50], 60))

    def test_target_not_found_in_gap(self):
        """Target falls between existing elements."""
        self.assertIsNone(interpolation_search([10, 20, 30, 40, 50], 25))

    def test_uniform_distribution(self):
        """Works well on uniformly distributed data."""
        array = list(range(0, 100, 10))  # [0, 10, 20, ..., 90]
        self.assertEqual(interpolation_search(array, 50), 5)

    def test_all_same_elements_found(self):
        """Array with all same elements, target matches."""
        self.assertEqual(interpolation_search([5, 5, 5, 5], 5), 0)

    def test_all_same_elements_not_found(self):
        """Array with all same elements, target doesn't match."""
        self.assertIsNone(interpolation_search([5, 5, 5, 5], 3))

    def test_two_elements_find_first(self):
        """Find the first element in a two-element array."""
        self.assertEqual(interpolation_search([10, 20], 10), 0)

    def test_two_elements_find_second(self):
        """Find the second element in a two-element array."""
        self.assertEqual(interpolation_search([10, 20], 20), 1)


if __name__ == "__main__":
    unittest.main()
